### PR TITLE
CAN Bus and Steering Display

### PR DESCRIPTION
**Pull request by ethanmpeterson**

### CODE REVIEW

File: templates/steering-display/platformio.ini [hunk 1]
```diff
@@ -26,8 +26,8 @@ board_build.f_cpu = 180000000L
 monitor_speed = 115200
 debug_tool = blackmagic
 upload_protocol = blackmagic
-upload_port = COM4 ; These ports will vary by host computer and platform
```
Comment (llim1) [inline templates/steering-display/platformio.ini]: Wonder if there is an easy way to prevent these lines from constantly being changed since ports are not static even on the same machine. Maybe we can just leave it commented by default?

File: templates/steering-display/src/main.cpp [hunk 2]
```diff
@@ -9,17 +9,26 @@
...
+
 void setup() {
     Serial2.begin(115200);
     pinMode(EVE_CS, OUTPUT);
     digitalWrite(EVE_CS, HIGH);
     pinMode(EVE_PDN, OUTPUT);
     digitalWrite(EVE_PDN, LOW);
 
+    pinMode(CAN_OE, OUTPUT);
+	digitalWrite(CAN_OE, LOW);
```
Comment (llim1) [inline templates/steering-display/src/main.cpp]: Just a formatting note. Not sure if this is just tabbed weird on the web?

File: templates/steering-display/src/main.cpp [hunk 3]
```diff
@@ -9,17 +9,26 @@
 #include <Arduino.h>
 #include <SPI.h>
 
+#include "can.hpp"
```
Comment (llim1) [inline templates/steering-display/src/main.cpp:7]: Not specific to CAN but got me thinking. It might be worthwhile organizationally speaking to add an `include` folder to this template and move the headers there.

File: Reference/steering-display/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+void setup() {
+    Serial2.begin(115200);
+    pinMode(EVE_CS, OUTPUT);
+    digitalWrite(EVE_CS, HIGH);
+    pinMode(EVE_PDN, OUTPUT);
+    digitalWrite(EVE_PDN, LOW);
+
+    pinMode(CAN_OE, OUTPUT);
+	digitalWrite(CAN_OE, LOW);
```
Comment (llim1) [inline Reference/steering-display/src/main.cpp]: Formatting here same as in the template.

File: Reference/steering-display/readme.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# EVE_Test_Arduino_PlatformIO Example
```
Comment (llim1) [inline Reference/steering-display/readme.md]: Modify readme to explain how this reference code works- what is different from the template?

**Review by llim1 — COMMENTED** [review]
Nothing major overall. Just some organizational stuff and formatting. 

I am a little concerned about CAL & CAN duplication across the repo having every project have its own src copy. We should dedicate a folder in the repo to CAL (and maybe CAN fw?) and have projects include headers as needed.

**Comment by ethanmpeterson** [comment]
> Nothing major overall. Just some organizational stuff and formatting.
> 
> I am a little concerned about CAL & CAN duplication across the repo having every project have its own src copy. We should dedicate a folder in the repo to CAL (and maybe CAN fw?) and have projects include headers as needed.

I agree with this, I just have not fixed that all up yet. Still a work in progress, we want everyone to reference a single shared copy of CAL.

File: Reference/steering-display/readme.md (hunk 5 above)
Comment (ethanmpeterson) [reply Reference/steering-display/readme.md]: TODO

File: templates/steering-display/src/main.cpp (hunk 3 above)
Comment (ethanmpeterson) [reply templates/steering-display/src/main.cpp:7]: Agreed. TODO

File: Reference/steering-display/src/main.cpp (hunk 4 above)
Comment (ethanmpeterson) [reply Reference/steering-display/src/main.cpp]: indent fix needed. TODO

File: templates/steering-display/src/main.cpp (hunk 2 above)
Comment (ethanmpeterson) [reply templates/steering-display/src/main.cpp]: probably messed up when copy pasted. TODO

File: templates/steering-display/platformio.ini (hunk 1 above)
Comment (ethanmpeterson) [reply templates/steering-display/platformio.ini]: Will comment out. No good way to assign static ports as far as I can tell outside of changing windows settings.

File: cal/README.md [hunk 6]
```diff
@@ -1,5 +1,22 @@
-# CAN Abstraction Layer (CAL)
+# CAL: CAN Abstraction Layer
```
Comment (ethanmpeterson) [inline cal/README.md]: @Jchisholm204 Please update this README to reflect the new directory structure of CAL. Most of this can go into the decoder subdirectory.

File: cal/decoder/library.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    "keywords": "CAN",
+    "authors":
+    [
+      {
+        "name": "Ethan Peterson",
+        "email": "ethan@petetech.net",
+        "url": "portfolio.petetech.net"
+      },
+      {
+        "name": "Andrew Smith",
```
Comment (ethanmpeterson) [inline cal/decoder/library.json]: @Jchisholm204 @mcassar4 Can you guys update these author fields? Please have everyone in the CAN group add their name! Great work on all of this!

File: cal/st-f4can/library.json [hunk 8]
```diff
@@ -0,0 +1,27 @@
+{
+    "name": "st-f4can",
+    "version": "1.0.0",
+    "description": "CAN controller asbtraction for CAN1 CAN2 peripherals on the STM32F446RE Microcontroller",
+    "keywords": "CAN",
+    "authors":
+    [
+      {
+        "name": "John Smith",
```
Comment (ethanmpeterson) [inline cal/st-f4can/library.json]: TODO for Ethan: Update Author fields. Remove unused dependency.

File: cal/st-f4can/library.json [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      {
+        "name": "Andrew Smith",
+        "email": "me@andrew-smith.com",
+        "url": "https://www.andrew-smith/contact",
+        "maintainer": true
+      }
+    ],
+    "license": "MIT",
+    "homepage": "qfsae.ca",
+    "dependencies": {
```
Comment (ethanmpeterson) [inline cal/st-f4can/library.json]: Remove this. No dependencies

File: steering/src/main.cpp [hunk 10]
```diff
@@ -1,16 +1,75 @@
...
+
+    if ((current_millis - previous_millis) > 3) {
+        // Execute touch display code every ~5 ms
+        previous_millis = current_millis;
+
+        TFT_touch();
+
+        display_delay++;
+        if (display_delay > 3) { // Execute display refresh every ~20ms
+            TFT_display((tft_callback)display_code);
```
Comment (ethanmpeterson) [inline steering/src/main.cpp:55]: TODO: Try and get this on a timer interrupt. Can be a separate PR.

File: steering/src/main.cpp [hunk 11]
```diff
@@ -1,16 +1,75 @@
...
+        }
+    }
+
+    uint8_t recv_ch = 1;
+    CAN_msg_t CAN_RX_msg;
+    if (CANMsgAvail(recv_ch)) {
```
Comment (ethanmpeterson) [inline steering/src/main.cpp:61]: TODO: Move this to interrupts. For separate PR.

File: steering/src/tft.c [hunk 12]
```diff
@@ -0,0 +1,198 @@
+/*
```
Comment (ethanmpeterson) [inline steering/src/tft.c:1]: Right now the TFT code is duplicated. Working on a platformio library fix for this. But might be helpful for those working on steering to merge as is. Refactor in the next PR

File: Reference/steering-display/readme.md (hunk 5 above)
Comment (ethanmpeterson) [reply Reference/steering-display/readme.md]: This was just a working copy. I have since deleted it from the branch. Not meant to be any different from what is in templates folder.

**Review by Jchisholm204 — APPROVED** [review]

File: steering/src/main.cpp (hunk 10 above)
Comment (llim1) [reply steering/src/main.cpp:55]: Let's make an issue for this and resolve for now?

File: steering/src/main.cpp (hunk 11 above)
Comment (llim1) [reply steering/src/main.cpp:61]: Same for this. Can be the same issue

**Comment by ethanmpeterson** [comment]
This branch has gone stale. Jacob and I will do a big refactor in another PR
